backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/monitoringannotation/aspect/MonitorAspect.java

```diff
@@ -0,0 +1,135 @@
+import org.adempiere.ad.window.api.IADWindowDAO;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.aspectj.lang.reflect.MethodSignature;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+import org.springframework.web.servlet.HandlerMapping;
+
+import javax.servlet.http.HttpServletRequest;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+@Aspect
+@Component
+public class MonitorAspect
+{
+	private final MicrometerPerformanceMonitoringService service;
+
+	public MonitorAspect(@NonNull final MicrometerPerformanceMonitoringService service)
+	{
+		this.service = service;
+	}
+
+	@Value("${performance.monitoring.enable:false}")
+	private Boolean perfMonEnvVar;
+
+	@Around("execution(* *(..)) && @annotation(de.metas.monitoringannotation.annotation.Monitor)")
+	public Object monitorMethod(ProceedingJoinPoint pjp) throws Throwable
+	{
+		final PerformanceMonitoringService.Metadata metadata;
+		final Callable callable = getCallableFromProceedingJoinPoint( pjp );
+
+		if(perfMonEnvVar = false)
+		{
+			callable.call();
+		}
+
+		Method method = ((MethodSignature) pjp.getSignature()).getMethod();
+		Monitor monitorAnnotation = method.getAnnotation(Monitor.class);
+
+		if(monitorAnnotation.type() == PerformanceMonitoringService.Type.REST_CONTROLLER_WITH_WINDOW_ID)
+		{
+			HttpServletRequest request = ((ServletRequestAttributes) RequestContextHolder.currentRequestAttributes()).getRequest();
+			Map pathVariables = (Map) request.getAttribute(HandlerMapping.URI_TEMPLATE_VARIABLES_ATTRIBUTE);
+			String windowId = (String)pathVariables.get("windowId");
+			String window = (Services.get(IADWindowDAO.class).retrieveWindowName(AdWindowId.ofRepoId(Integer.parseInt(windowId)))).getDefaultValue();
+
+			metadata = PerformanceMonitoringService.Metadata.builder()
+					.name(pjp.getTarget().getClass().getSimpleName())
+					.type(monitorAnnotation.type())
+					.action(method.getName())
+					.window(window + " (" + windowId + ")")
+					.build();
+		}
+		else
+		{
+			metadata = PerformanceMonitoringService.Metadata.builder()
+					.name(pjp.getTarget().getClass().getSimpleName())
+					.type(monitorAnnotation.type())
+					.action(method.getName())
+					.build();
+		}
+
+		return service.monitor(callable, metadata);
+
+	}
+
+	private Callable<Object> getCallableFromProceedingJoinPoint(final ProceedingJoinPoint pjp)
+	{
+		Callable<Object> callable = new Callable<Object>()
+		{
+
+			@Override
+			public Object call() throws Exception
+			{
+				try
+				{
+					return pjp.proceed();
+				}
+				catch (Exception e)
+				{
+					throw e;
+				}
+				catch (Throwable t)
+				{
+					throw new RuntimeException(t);
```

Comment:
why not `throw AdempiereException.wrapIfNeeded(t)`?
...or why catching non-Exception Throwables (i.e. Error) ? what's the reason?

Reply:
because of override i needed to add this catch